dds/DCPS/RTPS/Sedp.h

```diff
@@ -379,8 +386,55 @@ class Sedp : public DCPS::EndpointManager<ParticipantData_t> {
 #ifdef OPENDDS_SECURITY
   Writer participant_message_secure_writer_;
   Writer participant_stateless_message_writer_;
-  Writer participant_volatile_message_secure_writer_;
   Writer dcps_participant_secure_writer_;
+
+  /**
+   * Special Case for the Participant Volatile Message Secure Writer, which
+   * performs key exchange.
+   *
+   * When the auth initiator is much faster that the other side, the initiator
+   * might send the keys before the other side is ready to decrypt them. Given
+   * it is not durable, which means it will never ask for old data, the other
+   * side will never get the keys.
+   *
+   * The solution to this being used here is to wait until the first acknack is
+   * received, then send the keys again, just in case it didn't get them the
+   * first time.
+   */
+  class RepeatOnceWriter : public Writer {
+  public:
+    RepeatOnceWriter(const DCPS::RepoId& pub_id, Sedp& sedp);
+
+    /**
+     * If a remote participant's volatile reader acks us, resend all keys (User
+     * and Builtin) for a second time.
+     *
+     * Removes the participant's keys afterwards.
+     */
+    void acknowledged_by_reader(const DCPS::RepoId& rdr);
+
+    struct RemoteWriter {
+      DCPS::RepoId local_reader, remote_writer;
+      DDS::Security::DatareaderCryptoTokenSeq reader_tokens;
+    };
+    typedef OPENDDS_VECTOR(RemoteWriter) RemoteWriterVector;
+    typedef OPENDDS_MAP_CMP(
+      DCPS::RepoId, RemoteWriterVector, DCPS::GUID_tKeyLessThan) RemoteWriterVectors;
+    RemoteWriterVectors remote_writers_;
+
+    struct RemoteReader {
+      DCPS::RepoId local_writer, remote_reader;
+      DDS::Security::DatawriterCryptoTokenSeq writer_tokens;
+    };
+    typedef OPENDDS_VECTOR(RemoteReader) RemoteReaderVector;
+    typedef OPENDDS_MAP_CMP(
+      DCPS::RepoId, RemoteReaderVector, DCPS::GUID_tKeyLessThan) RemoteReaderVectors;
+    RemoteReaderVectors remote_readers_;
+
+    /// Lock for remote_readers_ and remote_writers_
+    ACE_Thread_Mutex lock_;
+
+  } participant_volatile_message_secure_writer_;
```

Comment:
This changes the declaration order of `participant_volatile_message_secure_writer_` relative to `dcps_participant_secure_writer_`, do constructor initializer lists need to change as well?

Reply:
Yes, I have already changed it.